gnrc_ndp_host: Initial import of host behavior of router discovery

authmillenon opened the pull request:
This imports [host behavior](https://tools.ietf.org/html/rfc4861#section-6.3) for [NDP router discovery](https://tools.ietf.org/html/rfc4861#section-6). This encompasses:

* building and sending of (semi-periodic) router solicitations (including all available options specified in [RFC 4861](https://tools.ietf.org/html/rfc4861))
* handling of router advertisements and their options

Depends on:
* ~~#3628: address-less link-layer handling for NDP~~
* ~~#3745: `gnrc_ipv6_netif` adaptations for router discovery~~ (merged)

authmillenon commented:
Rebased to current dependencies

OlegHahm commented on sys/include/net/gnrc/ipv6/netif.h:
```diff
@@ -264,6 +338,7 @@ void gnrc_ipv6_netif_remove(kernel_pid_t pid);
  */
 gnrc_ipv6_netif_t *gnrc_ipv6_netif_get(kernel_pid_t pid);
 
+#if (defined(MODULE_GNRC_NDP_ROUTER) || defined(MODULE_GNRC_SIXLOWPAN_ND_ROUTER))
```
What's the purpose of this ifdef?

authmillenon replied:
This was introduced in #3745 and proposed by @cgundogan (https://github.com/RIOT-OS/RIOT/pull/3745#discussion_r38347055). I liked the idea because it prevents that the function is called 'accidentally' ;-)

authmillenon replied:
But to answer your question in a direct manner: To generate NOP macros in the else case.

authmillenon commented:
Rebased to current #3628 to incorporate authmillenon/RIOT#17.

authmillenon commented:
Rebased to current master and dependencies.

authmillenon commented on sys/net/gnrc/network_layer/ndp/gnrc_ndp.c:
```diff
@@ -260,6 +275,102 @@ void gnrc_ndp_nbr_adv_handle(kernel_pid_t iface, gnrc_pktsnip_t *pkt,
+        /* calculate new random reachable time */
+        uint32_t reach_time = genrand_uint32_range(GNRC_NDP_MIN_RAND, GNRC_NDP_MAX_RAND);
+        if_entry->reach_time_base = byteorder_ntohl(rtr_adv->reach_time);
+        reach_time = (reach_time * if_entry->reach_time_base) / 10;
```
The 10 comes from the fact, that `GNRC_NDP_MIN_RAND` and `GNRC_NDP_MAX_RAND` are bigger by a factor of 10 than in the RFC (because they are fractions). Don't know how to express that, other than already documented here: https://github.com/RIOT-OS/RIOT/blob/master/sys/include/net/gnrc/ipv6/netif.h#L296

OlegHahm replied:
Actually, I understand neither this explanation nor the relation to the documentation of reach_time_base`.

authmillenon replied:
`reach_time` is calculated by taking `gnrc_ipv6_netif_t::reach_time_base` (set by the advertisement) and multiply it by a random value between 0.5 and 1.5 (https://tools.ietf.org/html/rfc4861#section-10). To avoid floating point numbers (and also have some variation) I used 5 and 15 as values for `GNRC_NDP_MIN_RAND` and `GNRC_NDP_MAX_RAND`. To get back to the original value I devided by ten (as I did on [interface initialization](https://github.com/RIOT-OS/RIOT/blob/master/sys/net/gnrc/network_layer/ndp/gnrc_ndp.c#L358))

OlegHahm replied:
Ok, so, how about a comment here saying something like "to avoid floating point number computations, the boundaries for the random values are multiplied by 10"?

kaspar030 replied:
> "to avoid floating point number computations, the boundaries for the random values are multiplied by 10"

"in order to avoid division, we divide." :)

authmillenon replied:
> "in order to avoid division, we divide." :)

I did not say to avoid division, I said to avoid going to floating point computations ;-) I basically introduce a low-profile fixed-point number calculation here, if you want to call it that ;-P

OlegHahm replied:
Just add a comment here and all is fine.

authmillenon commented:
Addressed comments in previous commits.

OlegHahm commented:
Please rebase

authmillenon commented:
Rebased to current master

authmillenon commented:
Rebased to current master and dependencies

OlegHahm commented on sys/net/gnrc/network_layer/ndp/gnrc_ndp.c:
```diff
@@ -275,6 +275,103 @@ void gnrc_ndp_nbr_adv_handle(kernel_pid_t iface, gnrc_pktsnip_t *pkt,
+    int sicmpv6_size = (int)icmpv6_size, l2src_len = 0;
+    uint16_t opt_offset = 0;
+    assert(if_entry != NULL);
+    if (!ipv6_addr_is_link_local(&ipv6->src) ||
```
Styling question, but for readability I prefer a newline between variable initializations and actual start of the functionality.

authmillenon replied:
I agree. Don't know, why I did it that way.

OlegHahm commented on sys/net/gnrc/network_layer/ndp/gnrc_ndp.c:
```diff
@@ -275,6 +275,105 @@ void gnrc_ndp_nbr_adv_handle(kernel_pid_t iface, gnrc_pktsnip_t *pkt,
+    /* set flags from message */
+    if_entry->flags &= ~GNRC_IPV6_NETIF_FLAGS_RTR_ADV_MASK;
+    if_entry->flags |= ((rtr_adv->flags << GNRC_IPV6_NETIF_FLAGS_RTR_ADV_POS) &
+                        (GNRC_IPV6_NETIF_FLAGS_RTR_ADV_MASK));
```
nitpicking, but I think you overdid with parens here. ;)

authmillenon replied:
Who are you and where is @OlegHahm?

OlegHahm replied:
(())

OlegHahm commented on sys/net/gnrc/network_layer/ndp/internal/gnrc_ndp_internal.c:
```diff
@@ -396,11 +397,79 @@ int gnrc_ndp_internal_tl2a_opt_handle(gnrc_pktsnip_t *pkt, ipv6_hdr_t *ipv6,
+                                     ndp_opt_pi_t *pi_opt)
+{
+    ipv6_addr_t *prefix;
+    gnrc_ipv6_netif_addr_t *a;
```
I would prefer `addr` or something more speaking.

OlegHahm commented:
apart from the minor comments: code is okay, test didn't break anything. ACK after addressing and squashing if Travis agrees, too.

authmillenon commented:
Addressed and squashed immediately. One minor thing I changed: I put the calculation and setting of reachable time related `gnrc_ipv6_netif_t` members into their own inline function and used that function on interface initialization as well.

authmillenon commented:
Thanks for the review! :-)
